src/tensor/cpu/device.rs

```diff
@@ -60,12 +60,8 @@ impl HasErr for Cpu {
     type Err = CpuError;
 }
 
-impl DeviceStorage for Cpu {
-    type Vec<E: Unit> = Vec<E>;
-
-    fn try_alloc_grad<E: Unit>(&self, other: &Self::Vec<E>) -> Result<Self::Vec<E>, Self::Err> {
-        self.try_alloc_zeros(other.len())
-    }
+impl<E: Unit> DeviceStorage<E> for Cpu {
+    type Storage = Vec<E>;
```

Comment:
I wonder if it's worthwhile to make this non-generic. I'm imagining:
```rust
impl DeviceStorage<f32> for Cpu { type Storage = Vec<f32> }
impl DeviceStorage<f64> for Cpu { type Storage = Vec<f64> }
impl DeviceStorage<bool> for Cpu { type Storage = Vec<bool> }
impl DeviceStorage<usize> for Cpu { type Storage = Vec<usize> }
...
impl DeviceStorage<Quantized<f32>> for Cpu { type Storage = Vec<QuantBlock> }
```

Reply:
& then we wouldn't need a separate QuantizedCpu device

Reply:
but we'd likely have to make the kernel impls non-generic too. Because I don't *think* you can have?
```rust
impl<E: Float> UnaryKernel<AbsKernelOp, E> for Cpu { }
impl UnaryKernel<AbsKernelOp, Quantized<f32>> for Cpu  { }
```

Reply:
Either way we'll need completely different impls for the kernels, iters, etc. No way around that unless we can match the access pattern of a `Vec`.

Reply:
Yep totally agree there, I like the approach of having `DeviceStorage<E>` - it fits well, and we gain a lot by this method. For instance we could represent sparse tensors by
```rust
struct Sparse<E>(E);

struct MaskedVec<E> {
    data: Vec<E>,
    mask: Vec<bool>,
}

impl<E> DeviceStorage<Sparse<E>> for Cpu {
    type Storage = MaskedVec<E>
}
```

which isn't possible with the current GAT approach.

Plus `Tensor<S, Sparse<E>, D>` or `Tensor<S, Quantized<E>, D>` looks pretty dang cool!